File: C/quicksort_binaria.c
// Pograma para ordenar un arreglo mediante el algoritmo de QuickSort y buscar un elemento mediante Binary Search

#include <stdio.h>

void fill_array(int arr[], int size)
{
    register int i;
    int element;

    for (i = 0; i < size; i++)
    {
        printf("\nIngrese el valor del elemento %d ", i);
        scanf("%d", &element);
        arr[i] = element;
    }
    printf("\n");
}

// Funcion para imprimir un arreglo

void print_array(int arr[], int size)
{
    register int i;

    printf("\nLos elementos del arreglo son: [ ");
    
    for (i = 0; i < size; i++)
    {
        printf("%d ", arr[i]);
    }
    
    printf("]\n");
}

// Funcion para particionar el arreglo

int partition(int arr[], int start, int end)
{
    register int j;
    int pivote, i, aux;
    pivote = arr[end];
    i = start - 1;

    for (j = start; j <= end - 1; j++)
    {
        if (arr[j] < pivote)
        {
            i++;
            aux = arr[i];
            arr[i] = arr[j];
            arr[j] = aux;
        }
    }
    i++;
    aux = arr[i];
    arr[i] = arr[end];
    arr[end] = aux;

    return i;
}

// Funcion para el ordenamiento por QuickSort

void quick_sort(int arr[], int start, int end)
{
    int pivote;

    if (start < end)
    {
        pivote = partition(arr, start, end);
        quick_sort(arr, start, pivote - 1);
        quick_sort(arr, pivote + 1, end);
    }
}

// Funcion para la binary search

int binary_search(int arr[], int size, int v)
{
    int start, end, mid;

    start = 0;
    end = size - 1;

    while (start <= end)
    {
        mid = (start + end) / 2;

        if (arr[mid] == v)
            return mid;
        else if (arr[mid] < v)
            start = mid + 1;
        else
            end = mid - 1;
    }

    return -1;
}

void main()
{
    int n, value, pos;

    printf("\nIngrese la cantidad de elementos del arreglo: ");
    scanf("%d", &n);

    int array[n];

    // Llamada a funcion para rellenar el arreglo

    fill_array(array, n);
    print_array(array, n);

    // Llamada a funcion para ordenar el arreglo mediante QuickSort

    printf("\nOrdenando mediante Quicksort...\n");
    quick_sort(array, 0, n - 1);
    print_array(array, n);

    // Busqueda de un elemento por binary search

    printf("Ingrese el valor del elemento a buscar: ");
    scanf("%d", &value);

    pos = binary_search(array, n, value);
    printf("\nLa posicion del elemento %d es: %d\n", value, pos);
}

File: Python/quicksort_binaria.py
# Programa para ordenar lista por quicksort y buscar un elemento por busqueda binaria

# Funcion para rellenar una lista

def fill_list(lista):
    length = int(input("\nIngrese la longitud de la lista: "))

    for i in range(0, length):
        element = int(input(f"\nIngresa el valor del elemento {i}: "))
        lista.append(element)

# Funcion para particionar una lista

def partition(lista, start, end):

    pivote = lista[end]
    i = start

    for j in range(start, end):
        if lista[j] < pivote:
            aux = lista[i]
            lista[i] = lista[j]
            lista[j] = aux
            i += 1

    aux = lista[i]
    lista[i] = lista[end]
    lista[end] = aux

    return i

# Funcion para ordenar por quicksort

def quick_sort(lista, start, end):

    if start < end:

        pivote = partition(lista, start, end)
        quick_sort(lista, start, pivote - 1)
        quick_sort(lista, pivote + 1, end)

# Funcion para la binary search

def binary_search(lista, size, v):

    start = 0
    end = size - 1
    mid = 0

    while start <= end:

        mid = (start + end) // 2

        if lista[mid] == v:
            return mid
        elif lista[mid] < v:
            start = mid + 1
        else:
            end = mid - 1

    return -1



# Programa main()

int_list = []

# Llamada a funcion para rellenar la lista

fill_list(int_list)
print(f"\nLa lista es: {int_list}")
list_size = len(int_list)

# Llamada a funcion para ordenar por quicksort

print(f"\nOrdenando por quicksort...\n")
quick_sort(int_list, 0, list_size - 1)
print(f"La lista ordenada es: {int_list}")

# Llamada a funcion para la busqueda binaria

value = int(input("Ingrese el valor del elemento a buscar: "))
pos = binary_search(int_list, list_size, value)
print(f"\nLa posicion del elemento {value} es: {pos}")